### PR TITLE
multi queue: Fix VMI creation in case non virtio interface present and multi vCPU

### PR DESCRIPTION
**Pull request by oshoval**
**What this PR does / why we need it**:
In case the following happen together:
1. Multi queue is enabled
2. There is more than one vCPU
3. There is at least one non virtio interface,

the VMI would fail to be created with the following error:
`"LibvirtError(Code=38, Domain=0, Message='Unable to create tap device tap0: Invalid argument')"`

Multi queue is only implemented for virtio interfaces.
The tap device which is created, didn't take into account the
interface model when calculating the number of queues.
This caused the problem for non virtio interfaces.
Fix it by taking into account the interface model.

`vmi.Spec.Domain.Devices.Rng` is removed from the test because it
is already included in `libvmi.NewFedora` which is used.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Note: this is not a regression of https://github.com/kubevirt/kubevirt/pull/8071

We might want to introduce NAD lib in order to reuse code.

**Release note**:
```release-note
None
```

### CODE REVIEW

**Review by maiqueb — CHANGES_REQUESTED** [review]
would you re-phrase your commit msg ?

The part I'm copying below is confusing.
> The reason is that the tap device which is created, didn't respect the
interface model, hence creating a multi queue tap device, whereas
the domain doesn't request multi queue for the corresponding interface
(because it does respect the model of the interface).

AFAIU, the reason is multi-queue is only implemented for `virtio` interfaces - as such,
it cannot be used on other iface types.

File: pkg/network/infraconfigurators/bridge.go
Comment (maiqueb) [inline pkg/network/infraconfigurators/bridge.go:140]: Can we stop adding attributes to this function ? 

How about you introduce a new wrapping struct - let say `tapDeviceInfo` - and group all this stuff in it ?

**Comment by oshoval** [comment]
> would you re-phrase your commit msg ?
> 
> The part I'm copying below is confusing.
> 
> > The reason is that the tap device which is created, didn't respect the
> > interface model, hence creating a multi queue tap device, whereas
> > the domain doesn't request multi queue for the corresponding interface
> > (because it does respect the model of the interface).
> 
> AFAIU, the reason is multi-queue is only implemented for `virtio` interfaces - as such, it cannot be used on other iface types.

Thats right, the domain xml did take into account the interface model,
but the tap device didn't (tried to say the same, will try to re-phrase it)

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go
Comment (maiqueb) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:645]: why are you removing the vmi admitting code ? Wouldn't it make sense to reject a request for multi-queue **if** an the corresponding VM features a non-virtio interface ?

**Review by maiqueb — CHANGES_REQUESTED** [review]
I know I've added a comment in a commit belonging to another PR, but I wonder ... Wouldn't it make more sense to just reject requests for multi-queue in VMIs that feature non-virtio interfaces ?

File: pkg/network/infraconfigurators/bridge.go
Comment (oshoval) [reply pkg/network/infraconfigurators/bridge.go:140]: I replaced one for one, so if refactor is needed lets do it please on a follow PR, not on this one

**Comment by maiqueb** [comment]
> > would you re-phrase your commit msg ?
> > The part I'm copying below is confusing.
> > > The reason is that the tap device which is created, didn't respect the
> > > interface model, hence creating a multi queue tap device, whereas
> > > the domain doesn't request multi queue for the corresponding interface
> > > (because it does respect the model of the interface).
> > 
> > 
> > AFAIU, the reason is multi-queue is only implemented for `virtio` interfaces - as such, it cannot be used on other iface types.
> 
> Thats right, the domain xml [did](https://github.com/kubevirt/kubevirt/pull/8219/commits/2fd1ec7e98692bba578d027666f8d05b8087a185#diff-49fbb923daeba230951affa978d449e7dcc8e775af82ef34d2994ac8545f29ecL74) take into account the interface model, but the tap device creation didn't (this what i tried to say, will try to re-phrase it)

Cool! The good thing is I never-the-less understood it. Thanks.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go
Comment (oshoval) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:645]: This what concluded in the long discussion here
https://github.com/kubevirt/kubevirt/issues/7671

That we relax it, it will be best effort
and present queueCount

I think it do belong to the mentioned PR instead please
https://github.com/kubevirt/kubevirt/pull/8071

Comment (maiqueb) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:645]: I think relaxing it is in fact making it less deterministic, which is only making it more confusing. 

But if you've agreed on it ... I guess I missed the opportunity to voice my concerns.

Comment (EdDev) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:645]: Relaxing means that we cannot maintain endless conditions of permutations.
The complexity is due to an API which supported defaults and the notion of parameters that apply on multiple interfaces.

After 2 or 3 incidents of bugs due to this matrix of permutations, we managed to convinced some of the maintainers that this is not maintainable. Thus, relaxing the strictness in favor of reflecting the actual state (number on queues).

My lesson out of this is basically to never try and abstract an API by introducing defaults and template parameters instead of individual ones (e.g. queue count per VMI and not per interface).

**Comment by oshoval** [comment]
Updated PR / commit message

**Comment by maiqueb** [comment]
> Updated PR / commit message

Thanks, it is now easier to understand.

**Comment by oshoval** [comment]
Ready to review, thanks
(rebased on what it was dependent so it is clean now)

File: tests/vmi_multiqueue_test.go
Comment (EdDev) [inline tests/vmi_multiqueue_test.go:184]: This is nice, consider having a following PR to extract it to `libnet` and use it by others tests as well.

File: tests/vmi_multiqueue_test.go
Comment (EdDev) [inline tests/vmi_multiqueue_test.go:73]: Please place such things closer to where they are used.

File: pkg/virt-launcher/virtwrap/converter/network.go
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/converter/network.go:201]: Can we get `"virtio"` from some central place (types?).

File: tests/vmi_multiqueue_test.go
Comment (oshoval) [reply tests/vmi_multiqueue_test.go:184]: Totally (was written on PR desc)
Follow PR sounds good

File: pkg/network/infraconfigurators/bridge.go
Comment (oshoval) [reply pkg/network/infraconfigurators/bridge.go:18]: I prefer that it will be done on a follow PR please.

**Comment by oshoval** [comment]
/test pull-kubevirtci-bump-kubevirt

Not related to this PR, just want a bump to be created

File: pkg/network/infraconfigurators/bridge.go
Comment (EdDev) [reply pkg/network/infraconfigurators/bridge.go:18]: Ok

File: pkg/virt-launcher/virtwrap/converter/network.go
Comment (oshoval) [reply pkg/virt-launcher/virtwrap/converter/network.go:201]: virtio is all over, 
added ModelVirtio which will be used for interface / device model virtio, and also for bus / disk / gfx model virtio
(didn't create define a specific for each entity)

will push soon

**Comment by oshoval** [comment]
Rebase

**Comment by oshoval** [comment]
Thanks
Addressed comments

**Comment by oshoval** [comment]
Fix compilation due to rebase

/test pull-kubevirt-e2e-k8s-1.24-sig-network

**Comment by oshoval** [comment]
Added missing empty line and missing Bazel file update

**Comment by oshoval** [comment]
rebased

**Comment by oshoval** [comment]
Rebased

File: tests/vmi_multiqueue_test.go
Comment (EdDev) [inline tests/vmi_multiqueue_test.go:75]: `context.TODO()` should be `context.Background()`. `TODO` is usually used when there is intention to replace it with something more appropriate in the future, but I think you do not have such an intention here.

File: staging/src/kubevirt.io/api/core/v1/types.go
Comment (EdDev) [inline staging/src/kubevirt.io/api/core/v1/types.go:820]: It seems that it is sometimes referred as `model`, sometime as `type` and sometimes `bus`.
How about just calling it `VirtIO`? Lets see what others think, this one is not a blocker for me.

**Review by EdDev — COMMENTED** [review]
Thank you.
I had only a few minor comments.

File: staging/src/kubevirt.io/api/core/v1/types.go
Comment (oshoval) [reply staging/src/kubevirt.io/api/core/v1/types.go:820]: Yep, i meant that it is the x model for example disk model / interface model (maybe even bus model)

https://github.com/kubevirt/kubevirt/pull/8219/commits/d9502000ecb2236c7b025f9b7e49817dc211912d#diff-de97cdb6be4b3bd722d6e7e208a39934957167715eef17717bf4c7009cc40cf5R171

does it make more sense, or you still prefer `VirtIO` only ?

File: pkg/virt-launcher/virtwrap/converter/network.go
Comment (oshoval) [reply pkg/virt-launcher/virtwrap/converter/network.go:215]: you mean to call it `isTrue` then ? (little weird imo, but whatever you prefer)

**Comment by oshoval** [comment]
Rebase

**Comment by oshoval** [comment]
Addressed comments

**Comment by oshoval** [comment]
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/8219/pull-kubevirt-e2e-k8s-1.23-sig-compute-nonroot/1568886254810959872

failure is due to this PR, need to fix it first

File: tests/vmi_multiqueue_test.go
Comment (AlonaKaplan) [inline tests/vmi_multiqueue_test.go:68]: I would create a new test "should be able to run a VM with multiqueue and a mix of virtio and non-virtio interfaces".
Now it is not clear why the second interface is added.

Comment (oshoval) [reply tests/vmi_multiqueue_test.go:68]: Note: no need a mix, an e1000 is enough

The conditions are:
1. Multi queue is enabled
2. There is more than one vCPU
3. There is at least one non virtio interface, (only one interface which is e1000 is good enough)

I just added another interface to an existing test. in order to spin less vms during e2e.

Does it change the picture?
Btw imho testing this corner case is an over kill

**Comment by oshoval** [comment]
Changes:
Spitted to two tests, following https://github.com/kubevirt/kubevirt/pull/8219#discussion_r974315876

**Comment by oshoval** [comment]
Rebased

File: tests/vmi_multiqueue_test.go
Comment (EdDev) [inline tests/vmi_multiqueue_test.go:87]: nit: Is the compiler requires this conversion here?
Also, one uses a constant (`numCpus`) and this one is not. This is strange.

Comment (EdDev) [reply tests/vmi_multiqueue_test.go:87]: I do not like the fact that we are wasting an e2e test for this check. Investing in an assertion makes sense, it is almost free, but having a dedicated test that is wasting around a minute for this is a waste of resources IMO.

As I noticed you received a comment to make it a separate test, I can only repeat my disagreement to it, nothing more.

**Review by EdDev — APPROVED** [review]
Thank you.

Except of the inline nits, I have a general nit about the comment title: They usually try to focus on the area the change took place.
In this case, I think you could provide a more accurate titles, e.g.:
- `network, multi-queue`
- `client/api, types`

I do not consider any as blockers, so you have my LGTM.

**Pull request merged**
